### PR TITLE
persist: add CI coverage for persist backed by s3

### LINKED ISSUE #10225: persistence,s3: "error: location "pstoev-persistence-test/LOCK" was already_locked" after double restart
### What version of Materialize are you using?

v0.18.1-dev (fc13fcb9f)

### How did you install Materialize?

Built from source

### What was the issue?

If the persistence test is patched as follows to use s3:

```
diff --git a/test/persistence/mzcompose.py b/test/persistence/mzcompose.py
index 40e22e28f..54561b654 100644
--- a/test/persistence/mzcompose.py
+++ b/test/persistence/mzcompose.py
@@ -31,6 +31,10 @@ mz_disable_user_indexes = Materialized(
     options=f"{mz_options} --disable-user-indexes",
 )
 
+mz_s3 = Materialized(
+    options=f"{mz_options} --persist-storage-enabled --persist-storage=s3://mtlz-test-persist-1d-lifecycle-delete/pstoev-persistence-test"
+)
+
 prerequisites = ["zookeeper", "kafka", "schema-registry"]
 
 SERVICES = [
@@ -45,7 +49,7 @@ td_test = os.environ.pop("TD_TEST", "*")
 
 
 def workflow_persistence(c: Composition) -> None:
-    for mz in [mz_default, mz_logical_compaction_window_off]:
+    for mz in [mz_s3, mz_default, mz_logical_compaction_window_off]:
         with c.override(mz):
             workflow_kafka_sources(c)
             workflow_user_tables(c)
```

Running it will cause Mz to fail post-restart:

```
materialized: persistence error: location "pstoev-persistence-test/LOCK" was already_locked:
fd40219b-5814-47ac-9a4e-b162ea78beac
materialized v0.18.1-dev (fc13fcb9f)
os: Ubuntu 20.4.0 [64-bit]
start time: 2022-01-25 11:39:23.513243968 UTC
num workers: 2
```

### Relevant log output

```source
2022-01-25T11:39:27.923317Z  INFO imds_load_region: aws_config::imds::region: loaded region from IMDS region=us-east-2
2022-01-25T11:39:27.948279Z  INFO send_operation{operation="GetObject" service="s3"}: aws_config::meta::credentials::chain: provider in chain did not provide credentials provider=Environment context=environment variable not set
2022-01-25T11:39:27.948337Z  INFO send_operation{operation="GetObject" service="s3"}: aws_config::meta::credentials::chain: provider in chain did not provide credentials provider=Profile context=No profiles were defined
2022-01-25T11:39:27.948347Z  INFO send_operation{operation="GetObject" service="s3"}: aws_config::meta::credentials::chain: provider in chain did not provide credentials provider=WebIdentityToken context=$AWS_WEB_IDENTITY_TOKEN_FILE was not set
2022-01-25T11:39:27.948357Z  INFO send_operation{operation="GetObject" service="s3"}: aws_config::meta::credentials::chain: provider in chain did not provide credentials provider=EcsContainer context=ECS provider not configured
2022-01-25T11:39:27.956293Z  INFO send_operation{operation="GetObject" service="s3"}: aws_config::meta::credentials::chain: loaded credentials provider=Ec2InstanceMetadata
materialized: persistence error: location "pstoev-persistence-test/LOCK" was already_locked:
fd40219b-5814-47ac-9a4e-b162ea78beac
materialized v0.18.1-dev (fc13fcb9f)
os: Ubuntu 20.4.0 [64-bit]
start time: 2022-01-25 11:39:23.513243968 UTC
num workers: 2
```

### PR DESCRIPTION
**Pull request by ruchirK**
Fixes #10225

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by ruchirK** [comment]
@philip-stoev this is one way we could run the existing mzcompose based tests on s3 while persist still maintains the restriction that only one cluster_id (unique entry per new catalog) can use a given "root location" in s3 at a time. This seems to run fine in scratch but fails to run in ci with this error:

```
materialized_1     | materialized: s3 get meta err: Error { code: "PermanentRedirect", message: "The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.", request_id: "E3GR48VJKZSDS7WX", s3_extended_request_id: "+gVeflAsdCdqStvJ5IXHhO2vAdZwakauX56hjDwqMihp+eDhQDXi7mX4OyzlxgXndotUVNQVUUM=" }
```

I don't know how to make progress with that - will pick this up tomorrow, unless perhaps you know a quick fix? Or perhaps, the s3 tests shouldn't run on every ci run and instead should only run on nightly?

**Review by philip-stoev — APPROVED** [review]
The Python part looks OK. Unfortunately I have no idea about the AWS/S3 error :-(

**Comment by ruchirK** [comment]
oh I'm an idiot - I force pushed to this branch which wiped out the other commit that Nikhil had pushed earlier

**Comment by ruchirK** [comment]
@philip-stoev this seems to be working now (got over the aws hump thanks to @benesch's help)

There's one last thing I find strange - we only have a failpoint for files at the moment. Therefore, the failpoint_set_sync.td fails as expected. That's totally fine. But, the failpoints test in `kafka-sources` succeeds. I think I know the root cause -- we insert so much data and then exit, so we most likely have not ingested the full dataset. Then on restart, we satisfy `failpoint-after.td`'s requirements -- we read in all the data, and we read a subset from kafka instead of from persist.

This seems flaky but I haven't actually observed it flake. I'm not aware of any good short term fixes. I'd love to avoid adding a s3blob failpoint in this pr if possible. Do you have any thoughts?

**Comment by philip-stoev** [comment]
@ruchirK we were in Barcelona so please give me until tomorrow to gather my thoughts on this and get back to you. I have an open task to add even more failpoints so I was hoping to have a universal (set of) .td files that could be run against any of N failpoints. So give me some time to think about how to structure this appropriately.

**Comment by ruchirK** [comment]
Aljoscha approved from Slack and this test ran 200 times without error so merging! TFTR!

**Review by philip-stoev — APPROVED** [review]
If the existing failpoint test passes with your PR, then I think it is good to go  -- please merge. As mentioned before I will deal with failpoints separately.

**Comment by ruchirK** [comment]
Ok cool. I'll rebase, check this test again ~100 times to make sure its not flaky, and add a todo that the failpoints test is file-specific, and then we can merge

**Comment by ruchirK** [comment]
Hmmm looks like the persist s3 test is now catching a bug:

```
materialized_1     | 2022-02-11T02:15:33.730943Z ERROR panic: assertion failed: `(left == right)`
materialized_1     |   left: `1644545727100`,
materialized_1     |  right: `1644545728500`: the upper of all involved streams must be the same
materialized_1     | thread: coordinator
materialized_1     | location: src/coord/src/persistcfg.rs:447:13
```

File: test/persistence/mzcompose.py
Comment (danhhz) [inline test/persistence/mzcompose.py:174]: how hard is it to separate failpoints out into its own workflow?

File: test/persistence/mzcompose.py
Comment (danhhz) [inline test/persistence/mzcompose.py:189]: nit: lets make this prefix named something more obviously connected to this composition. maybe `test-persistence-mzcompose`? another idea is `mzcompose-persistence-s3` to mirror the cli invocation of `./mzcompose --find persistence run s3`

File: test/persistence/mzcompose.py
Comment (ruchirK) [reply test/persistence/mzcompose.py:174]: Philip already did so in another PR that he opened today! I'll wait for that one to merge (and remove the todo) before merging this

File: test/persistence/mzcompose.py
Comment (ruchirK) [reply test/persistence/mzcompose.py:189]: hmm I like `mzcompose-persistence-s3`

File: test/persistence/mzcompose.py
Comment (danhhz) [reply test/persistence/mzcompose.py:174]: ha! I was looking at that next and came back here to amend, but you were too fast :)

**Review by danhhz — APPROVED** [review]
LGTM mod these two things (rebase and remove TODO, rename prefix)

File: test/persistence/mzcompose.py
Comment (philip-stoev) [reply test/persistence/mzcompose.py:174]: https://github.com/MaterializeInc/materialize/pull/10662 as been merged, so please proceed.

**Comment by ruchirK** [comment]
@danhhz I believe I have a commit that fixes a bug in the seal operator (and in so doing fixes the crash we were seeing with s3) but I have absolutely no idea:

- why this bugfix fixes the crash we were seeing
- why the crash we were seeing only presents itself on s3 and not on files

I've been tearing my hair out with this all day, and am going to take a break but perhaps its worth rubber ducking this tomorrow.

**Comment by danhhz** [comment]
> perhaps its worth rubber ducking this tomorrow

Happy to! Given that we essentially did our 1:1 a day early this week, let's use our 1:1 slot today for this

**Pull request closed without merging**
